### ISSUE TITLE
Reason for TristouNet triplet loss

### ISSUE THREAD

**Issue by evaldsurtans** [pyannote/pyannote-audio#120]
Thank you for the excellent work so far. I read many times your publications and pyannote code. Can you please explain what is the reason behind delta formulation

![image](https://user-images.githubusercontent.com/1777062/45413613-48e87e00-b682-11e8-8a58-af597be9e053.png)

My concerns:

1. If distances_between_positives = max_dist and distances_between_negatives = max_dist then delta = 0 
AND 
If distances_between_positives = 0 and distances_between_negatives = 0 then delta = 0 
This would lead to totally wrong results, correct? So alpha (margin) is always making it oscillate and get out of wrong loss, correct?

2. torch.clamp would make it oscillate forever around margin for negatives, correct? But we cannot guarantee that it will reach this point, because of my concern No.1, correct?

3. I see that you use also loss functions: softmargin and sigmoid, are they much better?

**Comment by evaldsurtans** [comment]
Maybe additional information could help to answer my question. 

Original triplet loss from "Distance metric learning for large margin nearest
neighbor classification, 2009"  
use 2 parts pull and push. But I see that all the latest works use only push part, including your implementation, **why?**  (see loss equations below from 2009)

Without first part loss does not converge for very long time.

![image](https://user-images.githubusercontent.com/1777062/45585897-a1738180-b8f5-11e8-90bd-2270dc5fd58a.png)

![image](https://user-images.githubusercontent.com/1777062/45585901-b2bc8e00-b8f5-11e8-8d7b-269f648d05de.png)

![image](https://user-images.githubusercontent.com/1777062/45585902-b6e8ab80-b8f5-11e8-89bd-8fdf2ff493a1.png)
